source/common/upstream/health_discovery_service.cc

```diff
@@ -54,12 +91,76 @@ void HdsDelegate::onReceiveInitialMetadata(Http::HeaderMapPtr&& metadata) {
+  ENVOY_LOG(debug, "New health check response message {} ", message->DebugString());
+  ASSERT(message);
+  for (int i = 0; i < message->health_check_size(); i++) {
+    // Create HdsCluster config
+    ENVOY_LOG(debug, "Creating HdsCluster config");
+    envoy::api::v2::core::BindConfig bind_config;
+
+    clusters_config_.emplace_back();
+    envoy::api::v2::Cluster& cluster_config = clusters_config_.back();
+
+    cluster_config.set_name(message->health_check(i).cluster_name());
+    cluster_config.mutable_connect_timeout()->set_seconds(ClusterTimeoutSeconds);
+    cluster_config.mutable_per_connection_buffer_limit_bytes()->set_value(
+        ClusterConnectionBufferLimitBytes);
+
+    // Add endpoints to cluster
+    const int localities_size = message->health_check(i).endpoints_size();
+    for (int j = 0; j < localities_size; j++) {
+      auto endpoints = message->health_check(i).endpoints(j);
+      for (int k = 0; k < endpoints.endpoints_size(); k++) {
+        cluster_config.add_hosts()->MergeFrom(endpoints.endpoints(k).address());
+      }
+    }
+
+    // TODO(lilika): Add support for optional per-endpoint health checks
+
+    // Add healthcheckers to cluster
+    auto health_checkers = message->health_check(i);
+    for (int j = 0; j < health_checkers.health_checks_size(); j++) {
+      cluster_config.add_health_checks()->MergeFrom(health_checkers.health_checks(j));
+    }
+
+    ENVOY_LOG(debug, "New HdsCluster config {} ", cluster_config.DebugString());
+
+    // Create HdsCluster
+    cluster_.reset(new HdsCluster(runtime_, cluster_config, bind_config, store_stats,
+                                  ssl_context_manager_, secret_manager_, false, info_factory_));
+
+    hds_clusters_.push_back(cluster_);
+
+    std::vector<Upstream::HealthCheckerSharedPtr> cluster_health_checkers;
+    for (int j = 0; j < message->health_check(i).health_checks_size(); j++) {
+      // Creating HealthCheckers
+      cluster_health_checkers.push_back(Upstream::HealthCheckerFactory::create(
+          cluster_config.health_checks(j), *hds_clusters_[i], runtime_, random_, dispatcher_));
+    }
+    health_checkers_ptr.push_back(cluster_health_checkers);
+  }
+}
+
+// TODO(lilika): Add support for subsequent HealthCheckSpecifier messages that
+// might modify the HdsClusters
 void HdsDelegate::onReceiveMessage(
     std::unique_ptr<envoy::service::discovery::v2::HealthCheckSpecifier>&& message) {
-  ENVOY_LOG(debug, "New health check response ", message->DebugString());
   stats_.requests_.inc();
-  stream_->sendMessage(health_check_request_, false);
-  stats_.responses_.inc();
+  ENVOY_LOG(debug, "New health check response message {} ", message->DebugString());
+
+  // Process the HealthCheckSpecifier message
+  processMessage(std::move(message));
+
+  // Initializing HealthCheckers
+  for (uint32_t i = 0; i < hds_clusters_.size(); i++) {
+    for (uint32_t j = 0; j < health_checkers_ptr[i].size(); j++) {
```

Comment:
Do we really benefit from having an nested structure here? Would `health_checker_` be better as just a flat vector? We basically just want to do "start all health checkers", the fact that they are grouped by cluster doesn't help much it seems.